### PR TITLE
Packaging

### PR DESCRIPTION
**Pull request by hellkite500**
Reorganize modules for packaging `bmi_sdk` and `bmi_pytorch` so they can be easily installed.  Will need to update the pyproject.toml for `bmi_pytorch` with a remote url for `bmi_sdk` once it has a stable url (after merge).

Tests all pass with pytest from any directory level.  Can start to add independent tests for the sdk modules at some point as well.

Will want to update the readme with some documentation on this once its ready, but wanted to get some feedback on the proposed packaging first.

@aaraney thoughts on this so far?

### CODE REVIEW

File: src/bmi_sdk/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+[project]
+name = "bmi_sdk"
+version = "0.0.0"
+dependencies = [
+  "bmipy",
+  "numpy"
+]
+
+#[tool.setuptools]
```
Comment (aaraney) [inline src/bmi_sdk/pyproject.toml:13]: You can remove this guy.

**Review by aaraney — APPROVED** [review]
Overall the package structure looks great! There was one commented out line that should be removed that looks like it got left in by accident. Nice stuff!

File: src/bmi_sdk/pyproject.toml (hunk 1 above)
Comment (hellkite500) [reply src/bmi_sdk/pyproject.toml:13]: Thanks for taking a look!  I'll get this dropped.

**Comment by hellkite500** [comment]
Hmm, well tests *were* running locally, but once I nuked the downloaded data dir which at one point was correctly created from `conftest.py`, I get the same failure locally that is showing up now on the load data test.  I'm really not sure why `pytest_sessionstart` doesn't seem to be executing now where it was when I worked up commit [afd98e1](https://github.com/hellkite500/bmi_pytorch/pull/15/commits/afd98e116b262e996dad7a9517bd5b9cd4f6fb60)

**Comment by taddyb** [comment]
I think I figured this out. The `conftest.py` is not running when it's defined inside multiple dirs. It has to be called from the root dir: https://stackoverflow.com/questions/34466027/what-is-conftest-py-for-in-pytest

> Test root path: This is a bit of a hidden feature. By defining conftest.py in your root path, you will have pytest recognizing your application modules without specifying PYTHONPATH. In the background, py.test modifies your sys.path by including all submodules which are found from the root path.

I moved the download_data part of the `conftest.py` to the root dir, and added something in the `README.md` in case someone doesn't know how to build the package for tests. Sorry for the multiple commits, I wanted to test the functionality works and passes tests.

**Comment by taddyb** [comment]
I recommend a new review of my changes since I had to add new files / a root `pyproject.toml` so our linter would ignore the `./tests/*` at the root level

**Review by taddyb — APPROVED** [review]
LGTM. My only thought is maybe to change the name of `src/bmi_pytorch/bmi_pytorch/tests/conftest.py` since there are only fixtures in here and nothing related to session initialization. I'm fine keeping it if we would like though as I probably have the least pytest experience here.

**Comment by hellkite500** [comment]
Moved from a session hook to a config hook for pytest so data download works for any pytest session initiated from any directory level.  Thanks @taddyb for getting to the root cause of that!

@aaraney mind taking one more quick look?

**Comment by aaraney** [comment]
> I think I figured this out. The `conftest.py` is not running when it's defined inside multiple dirs. It has to be called from the root dir: https://stackoverflow.com/questions/34466027/what-is-conftest-py-for-in-pytest
> 
> > Test root path: This is a bit of a hidden feature. By defining conftest.py in your root path, you will have pytest recognizing your application modules without specifying PYTHONPATH. In the background, py.test modifies your sys.path by including all submodules which are found from the root path.
> 
> I moved the download_data part of the `conftest.py` to the root dir, and added something in the `README.md` in case someone doesn't know how to build the package for tests. Sorry for the multiple commits, I wanted to test the functionality works and passes tests.

@taddyb, to be fair, `conftest.py` is a little esoteric and overloaded. [This](https://docs.pytest.org/en/7.1.x/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files) section of the `pytest` docs spell it out well:

> #### `conftest.py`: sharing fixtures across multiple files
> The conftest.py file serves as a means of providing fixtures for an entire directory. Fixtures defined in a conftest.py can be used by any test in that package without needing to import them (pytest will automatically discover them).
>
> You can have multiple nested directories/packages containing your tests, and each directory can have its own conftest.py with its own fixtures, adding on to the ones provided by the conftest.py files in parent directories.

So, having a `conftest.py` in the root directory will let you share fixtures across the project (and hook into `pytest`'s plugin hook system). As you alluded to, @taddyb, its usually best if you need to share fixtures across a single package's test suite to just have a `conftest.py` in that _package's_ test directory. Likewise, if you just need some fixtures for a single test _module_ (not shared between modules), just put them in that test file module.

Now in @hellkite500's case: 

> Hmm, well tests _were_ running locally, but once I nuked the downloaded data dir which at one point was correctly created from `conftest.py`, I get the same failure locally that is showing up now on the load data test. I'm really not sure why `pytest_sessionstart` doesn't seem to be executing now where it was when I worked up commit [afd98e1](https://github.com/hellkite500/bmi_pytorch/pull/15/commits/afd98e116b262e996dad7a9517bd5b9cd4f6fb60)

This is where `conftest.py` is a little overloaded and `pytest`'s default import behavior can really bite you (speaking from experience). Right, you can put fixtures in `conftest.py`, but you also add functionality to `pytest` via pytest's hook plugin system for [local plugins](https://docs.pytest.org/en/7.1.x/how-to/writing_plugins.html#conftest-py-local-per-directory-plugins). `pytest_sessionstart` is one of these hooks!

> Note:
> Some hooks should be implemented only in plugins or conftest.py files situated at the tests root directory due to how pytest discovers plugins during startup, see the documentation of each hook for details.

But, `pytest_sessionstart` is not one of those... So, this is where `pytest`'s default import / test collection behavior is wonky. TL;DR, for backwards compatibility, `pytest` is stuck with "`append`" imports, but in their docs they [recommend](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#tests-outside-application-code) that you use `importlib` instead. Ugh, more `pytest` things, you _can_ put this in `pyproject.toml` and that works well, when you just have a single package in the repo, but that is not our case... Instead, you can put this in a `pytest.ini` file in the root of the repo. This will fix that issue and others that will pop up in the future. This also means you can remove `__init__.py` files from `test`ing directories.

**Comment by aaraney** [comment]
I think in general, we can simply some of the package structure. Working on that now and will push to the branch.

**Comment by aaraney** [comment]
Packaging:

- flattened dir structure. moved packages back to root dir. I think this is easier to follow and more representative or what is going on.
- symlink LICENSE files to packages so they will be included during packaging.
- pyproject.toml's
    - set minimum python version to 3.8. `pytorch`'s current min supported version.
    - add authors
    - add licence
CI:

- install both packages from source
- echo dep versions -- should aid in debugging / reproducibility

**Comment by hellkite500** [comment]
@aaraney Thanks for cleaning up the data download. I was gonna get to that eventually 🤣.

I think this is in pretty good shape to help with development and usage! I'll leave it open for a bit to let @taddyb take a look, but excited to merge soon so I can knock off a few of the open issues that I didn't want to mess with too much till we were done rearranging package structure.

**Comment by hellkite500** [comment]
I spoke too soon, looks like we might need a minimum of python 3.9 based on current ci failures.

**Comment by hellkite500** [comment]
With 3.8 so near to EoL, I'm not too worried about supporting it, but if it's easy to fix for now we can.

**Comment by aaraney** [comment]
> With 3.8 so near to EoL, I'm not too worried about supporting it, but if it's easy to fix for now we can.

Yeah, that is generally my feeling too. I was mainly going off of `pytorch`'s minimum supported version. It was just a using `tuple` instead of `typing.Tuple`, so nothing crazy.

**Comment by aaraney** [comment]
All right, @hellkite500, got all the tests passing :tada:!

**Comment by aaraney** [comment]
In fixing that, I saw several unused imports in a few files. I am not going to fix those here to make this easier to review, but that should be addressed soon.

**Comment by taddyb** [comment]
I took a look at all of the commits and it all looks good to me 👍 

I wanted to put in writing the reason that the following line exists:
```
[tool.ruff]
exclude = [
    "./tests/*",
]
```

Excluding `tests/` from `ruff` is making sure the pre-commit ruff formatter/linter is not run on the tests folder as almost always throws some sort of usage error with `import pytest` or some fixture that is imported but only called as an arg. 

If there is a better way to do this, feel free to change this

**Pull request merged**